### PR TITLE
LibWeb: Add support for DOM::Document to XHR::send()

### PR DESCRIPTION
**Pull request by kennethmyhra**
This patch adds support for posting a `DOM::Document` using `XHR::send()`.

### CODE REVIEW

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp [hunk 1]
```diff
@@ -433,12 +433,19 @@ WebIDL::ExceptionOr<void> XMLHttpRequest::send(Optional<Fetch::XMLHttpRequestBod
     auto request = LoadRequest::create_for_url_on_page(request_url, m_window->page());
     request.set_method(m_method);
     if (body_with_type.has_value()) {
-        TRY_OR_RETURN_OOM(realm, body_with_type->body.source().visit([&](ByteBuffer const& buffer) -> ErrorOr<void> {
+        TRY(body_with_type->body.source().visit(
+            [&](ByteBuffer const& buffer) -> WebIDL::ExceptionOr<void> {
                 request.set_body(buffer);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:447]: Isn't this also fallible? ByteBuffer has a silly copy ctor, hiding all the implicit copies of potentially huge buffers...

**Review by linusg — COMMENTED** [review]
First commit message should probably mention XHR :)

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp (hunk 1 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:447]: Would it help if there was a `try_` implementation of the `set_body()` method. That's the only way I can think of at the moment that can can handle the fallible case.
```cpp
ErrorOr<void> LoadRequest::try_set_body(ByteBuffer const& body) { m_body = TRY(ByteBuffer::copy(body.bytes())); }
```

Comment (kennethmyhra) [reply Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp]: Let's try this

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp [hunk 2]
```diff
@@ -130,6 +130,9 @@ CppType idl_type_name_to_cpp_type(Type const& type, Interface const& interface)
     if (type.name() == "BufferSource")
         return { .name = "JS::Handle<JS::Object>", .sequence_storage_type = SequenceStorageType::MarkedVector };
 
+    if (type.name() == "XMLHttpRequestBodyInit")
+        return { .name = "Fetch::XMLHttpRequestBodyInit", .sequence_storage_type = SequenceStorageType::MarkedVector };
```
Comment (linusg) [inline Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp:134]: I thought I made a comment here, but it seems to be lost: why is this needed? :thinking:

File: Userland/Libraries/LibWeb/Loader/LoadRequest.h [hunk 3]
```diff
@@ -35,6 +35,18 @@ class LoadRequest {
...
+    {
+        m_body = TRY(ByteBuffer::copy(body.bytes()));
+        return {};
+    }
+
+    ErrorOr<void> try_set_body(ReadonlyBytes const& bytes)
+    {
+        m_body = TRY(ByteBuffer::copy(bytes));
+        return {};
+    }
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Loader/LoadRequest.h:48]: This doesn't seem ideal - `set_body()` takes a reference and always copies, so it should just take by value and move. Surely there aren't too many instances to update? :sweat_smile:

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp (hunk 2 above)
Comment (kennethmyhra) [reply Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp:134]: Our IDLGenerator needs it after changing the signature of `XHR::send()` method from `Optional<Fetch::XMLHttpRequestBodyInit>` to `Optional<Variant<JS::Handle<DOM::Document>, Fetch::XMLHttpRequestBodyInit>>`.

Without it we get the `Optional<Variant<JS::Handle<DOM::Document>, Fetch::XMLHttpRequestBodyInit>>` and hits a TODO.

File: Userland/Libraries/LibWeb/Loader/LoadRequest.h (hunk 3 above)
Comment (kennethmyhra) [reply Userland/Libraries/LibWeb/Loader/LoadRequest.h]: I am on thin ice when it comes to move. Did some read up on and I think I got how it works now 😅

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp (hunk 2 above)
Comment (linusg) [reply Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp:134]: Kinda weird that it can figure out "nullable type" but needs help for "nullable type that's either this or that", but let's fix that some other day :)

File: Userland/Libraries/LibWeb/Loader/LoadRequest.h [hunk 4]
```diff
@@ -33,7 +33,7 @@ class LoadRequest {
     void set_method(String const& method) { m_method = method; }
 
     ByteBuffer const& body() const { return m_body; }
-    void set_body(ByteBuffer const& body) { m_body = body; }
+    void set_body(ByteBuffer const body) { m_body = move(body); }
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Loader/LoadRequest.h:36]:
```suggestion
    void set_body(ByteBuffer body) { m_body = move(body); }
```

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp [hunk 5]
```diff
@@ -432,17 +439,30 @@ WebIDL::ExceptionOr<void> XMLHttpRequest::send(Optional<Fetch::XMLHttpRequestBod
...
-                return {}; }, [&](JS::Handle<FileAPI::Blob> const& blob) -> ErrorOr<void> {
-                auto byte_buffer = TRY(ByteBuffer::copy(blob->bytes()));
-                request.set_body(byte_buffer);
-                return {}; }, [](auto&) -> ErrorOr<void> { return {}; }));
-        if (body_with_type->type.has_value()) {
-            // If type is non-null and this’s headers’s header list does not contain `Content-Type`, then append (`Content-Type`, type) to this’s headers.
-            if (!m_request_headers.contains("Content-Type"sv))
-                request.set_header("Content-Type", String { body_with_type->type->span() });
+    if (serialized_document.has_value()) {
+        request.set_body(move(serialized_document.value()));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:443]:
```suggestion
        request.set_body(serialized_document.release_value());
```

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp [hunk 6]
```diff
@@ -432,17 +439,30 @@ WebIDL::ExceptionOr<void> XMLHttpRequest::send(Optional<Fetch::XMLHttpRequestBod
...
+    } else if (body_with_type.has_value()) {
+        TRY(body_with_type->body.source().visit(
+            [&](ByteBuffer const& buffer) -> WebIDL::ExceptionOr<void> {
+                request.set_body(move(buffer));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:447]: This still ends up copying, you can't `move()` a reference like that. The copy ctor really needs to go.

```suggestion
                auto byte_buffer = TRY_OR_RETURN_OOM(realm, ByteBuffer::copy(buffer));
                request.set_body(move(byte_buffer));
```

**Review by linusg — COMMENTED** [review]
Almost LGTM now :)

**Review by linusg — APPROVED** [review]

**Comment by kennethmyhra** [comment]
Dammit, I need a better understanding of `move`. As always, thanks for the input and patience, Linus
